web_src/js/components/file_tree.ts

```diff
@@ -0,0 +1,90 @@
+export type FileStatus = 'added' | 'modified' | 'deleted' | 'renamed' | 'copied' | 'typechange';
+
+export type File = {
+  Name: string;
+  NameHash: string;
+  Status: FileStatus;
+  IsViewed: boolean;
+  IsSubmodule: boolean;
+}
+
+type DirItem = {
+    isFile: false;
+    name: string;
+    path: string;
+
+    // we disable eslint here because we have a recursive type definition
+    // eslint-disable-next-line no-use-before-define
+    children: Item[];
+}
+
+type FileItem = {
+    isFile: true;
+    name: string;
+    path: string;
+    file: File;
+}
+
+export type Item = DirItem | FileItem;
+
+export function pathListToTree(fileEntries: File[]): Item[] {
+  const pathToItem = new Map<string, DirItem>();
+
+    // init root node
+  const root: DirItem = {name: '', path: '', isFile: false, children: []};
+  pathToItem.set('', root);
+
+  for (const fileEntry of fileEntries) {
+    const [parentPath, fileName] = splitPathLast(fileEntry.Name);
+
+    let parentItem = pathToItem.get(parentPath);
+    if (!parentItem) {
+      parentItem = constructParents(pathToItem, parentPath);
+    }
+
+    const fileItem: FileItem = {name: fileName, path: fileEntry.Name, isFile: true, file: fileEntry};
+
+    parentItem.children.push(fileItem);
+  }
+
+  return root.children;
+}
+
+function constructParents(pathToItem: Map<string, DirItem>, dirPath: string): DirItem {
+  const [dirParentPath, dirName] = splitPathLast(dirPath);
+
+  let parentItem = pathToItem.get(dirParentPath);
+  if (!parentItem) {
+        // if the parent node does not exist, create it
+    parentItem = constructParents(pathToItem, dirParentPath);
+  }
+
+  const dirItem: DirItem = {name: dirName, path: dirPath, isFile: false, children: []};
+  parentItem.children.push(dirItem);
+  pathToItem.set(dirPath, dirItem);
+
+  return dirItem;
+}
+
+function splitPathLast(path: string): [string, string] {
+  const lastSlash = path.lastIndexOf('/');
+  return [path.substring(0, lastSlash), path.substring(lastSlash + 1)];
+}
```

Comment:
This can be replaced by `basename` and `dirname` functions. The former already exists, the latter is easy to add:

https://github.com/go-gitea/gitea/blob/38ccc8e3e48e9b15f175287c0721ce0ff9892e5a/web_src/js/utils.ts#L5